### DISCUSSION TITLE
Preparing everything at initialization, nothing at runtime

### DISCUSSION THREAD

**Discussion by perwah** [jdbi/jdbi#2447] [Ideas]
I found myself debugging into JDBI when developing my project that uses JDBI SQL Objects, and I find that all the complex mapping, finding constructors, fields, mappers and argument-functions etc. is done in runtime when my application wants to read/write to the database. That's a lot of code that has to run every time I call my repository-methods, that always result in the same execution paths.

Wouldn't it be a nice performance gain to look-up and prepare all those things at initialization, resulting in ready-to-run chains of lambdas for each data class, that then just execute when it is time to access the database? This would _**eliminate the execution of hundreds of lines of code**_ at every invocation.

I spent an afternoon making a project (as a POC) that can introspect domain-object classes and prepare ready-to-run lambdas that create objects when reading, and takes data from objects when writing - all that has to be done when reading or writing an object is to invoke the appropriate lambda-chain, that already knows what getter-methods to read from or constructors to call and populate its parameters.

This way, the code that runs when writing an object to the database looks something like this:

```
public <T> boolean execute(PreparedStatement preparedStatement, T dataObject) throws SQLException {
    try {
        for (final var sqlValueSetter : sqlValueSetters) {
            sqlValueSetter.set(preparedStatement, dataObject);
        }
        return statement.execute();
    } catch (InvocationTargetException | IllegalAccessException e) {
        throw new RuntimeException(e);
    }
}

```
and that's about all that needs to be done by the framework when performing an SQL update. All the sqlValueSetters are lambdas that have been prepared in advance at initialization, that get values from the pre-looked-up getters of the dataObject, and write them to the pre-determined parameters of the statement. No runtime introspection or looking for mappers, just execute.

This means that for instance ConstructorMapper et.al. would be run lazily at first execution and set everything up.

**Comment by perwah** [comment]
Well, I wrote a test for my own little software that inserted and then read back 100 000 data objects, and then wrote a similar test for the JDBI framework (both with H2 in-mem database), and the times were pretty equal - about 10% faster for my code, but then it lacks a lot of features (like named parameters, and all the rest of the JDBI features.)

I think it is difficult to make it any faster than what I did, considering that at read/write time, there's no extra code running but a tiny loop reading/writing values from/to the resultSet/statement. For every object read/written, this is the only code run (with no extra overhead per data object, just get the next one in the list):

For every object in list to be saved on a batch insert (with list of objects):
```
for (final var sqlValueSetter : sqlValueSetters) {
    sqlValueSetter.set(preparedStatement, dataObject);
}
```
where `sqlValueSetter` is pre-created to get its value by invoking a getter-Method (using `FieldValueSupplier` below, just a lambda doing `Method::invoke` already obtained by reflection at initialization), and call the appropriate set-method in the prepared statement with a predetermined parameter-index:
```
public SqlValueSetter<Integer> getSqlValueSetter(final int paramIndex, final FieldValueSupplier<Integer> dataSupplier) {
    return (statement, dataObject) ->
        statement.setInt(paramIndex, dataSupplier.get(dataObject));
}

```
For every row in resultSet on a select:
```
for (final var sqlValueGetter : sqlValueGetters) {
    dataObjectValues.add(sqlValueGetter.get(resultSet));
}
return dataObjectFactory.newInstance(dataObjectValues);
```

JDBI test case used (adapted from `FiveMinuteTourTest.java`), runs in 900 ms:
```
    @Test
    public void positionalParameters() {
        int rowCount = 100_000;
        handle.registerRowMapper(new ContactMapper());

        final var list = new ArrayList<Contact>(rowCount);
        for (int i = 1; i <= rowCount; i++) {
            list.add(new Contact(i, "Chuck-" + 1));
        }
        PreparedBatch preparedBatch = handle.prepareBatch("insert into contacts (id, name) values (?, ?)");
        for (final var contact : list) {
            preparedBatch
                .bind(0, contact.id)
                .bind(1, contact.name)
                .add();
        }
        preparedBatch.execute();

        final var contacts = handle.createQuery("select * from contacts")
            .mapTo(Contact.class)
            .list();

        assertThat(contacts).hasSize(rowCount);
    }
```

My own software testcase used, runs in 800 ms:
```
  @Test
  public void test() throws SQLException {
    final int rowCount = 100_000;
    final var list = new ArrayList<TestObject>(rowCount);
    for (int i = 1; i <= rowCount; i++) {
      list.add(new TestObject(i, "Perkules" + i));
    }

    sqlMap.update(list, "insert into test_object (id, name) values (?, ?)");

    final var actual = sqlMap.select(TestObject.class, "select * from test_object");

    assertThat(actual).hasSize(rowCount);
  }
```
and after warmup (duplicating the test-method and running the same thing again), the time difference drops to below 5%. Perhaps more could be done using something like javassist, that would maybe be an interesting endeavour.

So to conclude, in this simple scenario I don't really see any significant performance gains to be had. I had worries, but as I tried to implement something that does as little as possible at runtime, I find that (for this simple use-case of saving 100k objects and then reading them back) there's really not much to be done, and I can continue to use JDBI without worrying about performance.

**Comment by stevenschlansker** [comment]
Hi @perwah , thank you for your interest in this topic. I agree, Jdbi should bind as much as it can ahead of time, although usually the observation is that in most cases the majority of time is spent in the database anyway.

We made some significant performance improvements in the upcoming `3.40.0` (a pre-release of which is available now as `3.40.0-rc1`).

If you do run into any specific performance problems in the Jdbi code, please do report them. But we would like to see an actual use case that it matters for.

For running performance tests, I strongly recommend `jmh`: https://github.com/openjdk/jmh/blob/master/README.md as just running a unit test does not control for background tasks like code compilation.

I will mark this as answered, since it sounds like you don't have a specific concern at this time, but please report back if you do run into bottlenecks.

**Reply by perwah** [reply]
When I optimized my test cases by preparing the data lists ahead of the tests, the difference grew to about 30%.
It was more of a fun experiment, as I saw an "optimal path" in my head, and couldn't stop myself from exploring it.

An interesting note is that using javassist to generate code that doesn't loop through value-setters, but just has lines of code directly setting the values, didn't improve performance *at all* for objects with 6 fields.

Javassist generated code example:
```
public void map(java.sql.PreparedStatement ps, java.lang.Object object) {
  try {
    org.sqlmap.BasicTest$TestObject dataObject = (org.sqlmap.BasicTest$TestObject) object;
    ps.setInt(1, dataObject.getId());
    ps.setString(2, dataObject.getName());
    ps.setString(3, dataObject.getSurname());
    ps.setInt(4, dataObject.getCost());
    ps.setString(5, dataObject.getAddress());
    ps.setString(6, dataObject.getComment());

  } catch (java.sql.SQLException e) {
    throw new org.sqlmap.SqlMapException("Couldn't write object of type " + object.getClass(), e);
  }
}
```
performs the same as
```
public void map(java.sql.PreparedStatement ps, java.lang.Object object) {
  try {
    for (final var sqlValueSetter : sqlValueSetters) {
      sqlValueSetter.set(ps, object);
    }
  } catch (java.sql.SQLException e) {
    throw new org.sqlmap.SqlMapException("Couldn't write object of type " + object.getClass(), e);
  }
}
```

The basis of my thoughts were "do as little as absolutely possible when interfacing with the database", so I worked backwards from that, to make something that I believe is very difficult to improve on, from a first principles point of view. Then perhaps with adding features and plugins, things are bound to deviate from the optimal path.

But all in all, I think it is important to work from the assumption that the "inner loop" should be well prepared, and the active code should just run through the loop with as much as possible prepared beforehand.

Like prepared statements. It may be beneficial to keep them around as long as the connection is open. Thus it may be beneficial to keep the connection and share it without closing. Etc and etc.

**Reply by perwah** [reply]
Just to comment on the statement that "in most cases the majority of time is spent in the database anyway":

For a CPU, the time spent waiting for the database is time spent just waiting (the thread doing nothing), giving other code time to run. If the code that needs to be run for database access is minimized, the CPU can do more things for more clients.
